### ISSUE TITLE
Xtend switch - implicit typecast problem

### ISSUE THREAD

**Issue by miklossy** [eclipse/xtext-xtend#435]
Given is the following project with two Xtend classes:
![screenshot](https://user-images.githubusercontent.com/20393472/39232534-f11fb63e-486d-11e8-8985-9fa56588d9e2.png)

There should be no error on line 14 and line 15 because of the implicit cast.
Hint: making the list variable public in the class Foo the errors are gone and the implicit cast works like a charm!

**Comment by cdietrich** [comment]
can you please attach code snippets too?

**Comment by cdietrich** [comment]
@szarnekow i am not sure if typeguards with generics make sense at all

**Comment by miklossy** [comment]
```
package foo

import java.util.List


class Foo {
	/*public*/ List<String> list

	new(List<String> list) {
		this.list = list
	}

	def getList(){
		list
	}

}
```


```
package foo

import java.util.ArrayList
import java.util.LinkedList

class Main {
	
	def static void main(String[] args) {
		new Foo(newArrayList).m
	}
	
	private def static m(Foo it){
		switch list {
			ArrayList<String>:  m1(list)
			LinkedList<String>: m2(list)
		}		
	}	
	
	private def static m1(ArrayList<String> list){
		println("ArrayList")
	}
	
	private def static m2(LinkedList<String> list){
		println("LinkedList")
	}
}
```

**Comment by miklossy** [comment]
The problem occurs on an examples that do not use generics at all. I created an another example without generics:
![screenshot2](https://user-images.githubusercontent.com/20393472/39234385-66775324-4873-11e8-9a2a-52f6ce207f82.png)

```
package foo

class Foo {
	/*public*/ A a

	new(A list) {
		this.a = a
	}

	def getA(){
		a
	}

}

class A {}
class B extends A{}
class C extends A{}
```

```
package foo

class Main {
	
	def static void main(String[] args) {
		new Foo(new B).m
	}
	
	private def static m(Foo it){
		switch a {
			B: m1(a /*as B*/)
			C: m2(a /*as C*/)
		}
	}	
	
	private def static m1(B b){
		println("B")
	}
	
	private def static m2(C c){
		println("C")
	}
}
```

**Comment by szarnekow** [comment]
I suspect you are calling the getter instead of accessing the field. Calling a method may yield different results when called multiple times. Thus casting is not supposed to work for that case.

**Comment by cdietrich** [comment]
hmmmm

```
class A {}
class B extends A {}
class C extends A {}
class D {
	A a
	def getA() {a}
}
 class E {
	def m1(B b) {}
	def m2(C c) {}
	def test(A a) {
		switch a {
			B: m1(a)
			C: m2(a)
		}
	}
	def test(D it) {
		switch a {
			B: m1(a)
			C: m2(a)
		}
	}
	def test2(D it) {
		val a = it.a
		switch a {
			B: m1(a)
			C: m2(a)
		}
	}
}
```

=> looks like the problem is switching on an expression and not a variable

**Comment by miklossy** [comment]
The example of @cdietrich represents the problem very well.

**Comment by szarnekow** [comment]
If you annotate your getter as `@Pure`, it should work, too.

**Comment by szarnekow** [comment]
No, it doesn't :(

**Comment by szarnekow** [comment]
In any case, this is intentional behavior right now and we'd need to figure out, if we want to widen the scope of that feature.

Please keep in mind, that you could always assign a local var within the switch expression:

```
      def test(D it) {
		switch a: a {
			B: m1(a)
			C: m2(a)
		}
	}
```

**Comment by miklossy** [comment]
I understand that it is the intended behaviour, since it is safer not to allow automatic type cast when switching over an expression, only allow it when switching over a variable. I like the suggestion/workaround of @szarnekow above.

I would suggest to describe this intended behaviour on the [Xtend documentation](http://www.eclipse.org/xtend/documentation/203_xtend_expressions.html#switch-expression) and close this ticket after that.

**Comment by miklossy** [comment]
I found the following description in the [Xbase Switch Expression documentation](https://www.eclipse.org/Xtext/documentation/305_xbase.html) 

> if the switch expression contains a variable declaration using the syntax known from for loops, the value is bound to the given name.

Fixed in 2.22.
